ocaml/xapi/xapi_observer.ml

```diff
@@ -110,14 +110,139 @@ module Observer : ObserverInterface = struct
+      ^ String.concat ","
+          (List.map (fun (k, v) -> k ^ "=" ^ v) config.otel_resource_attributes)
+    ; "OTEL_EXPORTER_ZIPKIN_ENDPOINTS="
+      ^ String.concat "," config.otel_exporter_zipkin_endpoints
+    ; "XS_EXPORTER_BUGTOOL_ENDPOINT="
+      ^ Option.value ~default:"" config.xs_exporter_bugtool_endpoint
+    ]
+
+  let remove_config ~uuid =
+    Xapi_stdext_unix.Unixext.unlink_safe
+      (config_root
+      ^ Component.component_name
+      ^ "/enabled/"
+      ^ uuid
+      ^ ".observer.conf"
+      )
+
+  let update_config ~__context ~observer ~config_root ~uuid =
+    if Db.Observer.get_enabled ~__context ~self:observer then (
+      let observer_config =
+        ObserverConfig.config_of_observer ~__context
+          ~component:Component.component_name ~observer
+      in
+      let dir_name = config_root ^ Component.component_name ^ "/enabled/" in
+      Xapi_stdext_unix.Unixext.mkdir_rec dir_name 0o755 ;
+      let file_name = dir_name ^ uuid ^ ".observer.conf" in
+      Xapi_stdext_unix.Unixext.write_string_to_file file_name
+        (String.concat "\n" (env_vars_of_config observer_config))
+    ) else
+      remove_config ~uuid
+
+  let update_all_configs ~__context ~observer_all =
+    List.iter
+      (fun observer ->
+        let uuid = Db.Observer.get_uuid ~__context ~self:observer in
+        update_config ~__context ~observer ~config_root ~uuid
+      )
+      observer_all
+
+  let create ~__context ~uuid ~name_label ~attributes ~endpoints ~enabled =
+    let observer = Db.Observer.get_by_uuid ~__context ~uuid in
+    update_config ~__context ~observer ~config_root ~uuid
+
+  let destroy ~__context ~uuid = remove_config ~uuid
+
+  let set_enabled ~__context ~uuid ~enabled =
+    let observer = Db.Observer.get_by_uuid ~__context ~uuid in
+    update_config ~__context ~observer ~config_root ~uuid
+
+  let set_attributes ~__context ~uuid ~attributes =
+    let observer = Db.Observer.get_by_uuid ~__context ~uuid in
+    update_config ~__context ~observer ~config_root ~uuid
+
+  let set_endpoints ~__context ~uuid ~endpoints =
+    let observer = Db.Observer.get_by_uuid ~__context ~uuid in
+    update_config ~__context ~observer ~config_root ~uuid
+
+  let init ~__context =
+    let observer_all = Db.Observer.get_all ~__context in
+    update_all_configs ~__context ~observer_all
+
+  let set_trace_log_dir ~__context ~dir =
+    let observer_all = Db.Observer.get_all ~__context in
+    update_all_configs ~__context ~observer_all
+
+  let set_export_interval ~__context ~interval = ()
+
+  let set_max_spans ~__context ~spans = ()
+
+  let set_max_traces ~__context ~traces = ()
+
+  let set_max_file_size ~__context ~file_size = ()
+
+  let set_host_id ~__context ~host_id = ()
+
+  let set_compress_tracing_files ~__context ~enabled = ()
+end
+
+module SMObserverConfig = Dom0ObserverConfig (struct
+  let component_name = "smapi"
+end)
+
+let supported_components = ["xapi"; "xenopsd"; "smapi"]
```

Comment:
we have now 3 supported components, these are strings. Do we need a variant instead here to avoid typos?

Reply:
Four: see Vincent's PR, where I suggested the same :)

Reply:
Also, I think the same idea applies to this string "bugtool" in these two places:
* https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_observer.ml#L163
* https://github.com/xapi-project/xen-api/pull/5266/commits/aa2d80506f23c04e7e51252dab74e455fa99e628#diff-a31ed9ca9a7a8e83afdb6bee0ad80e86361374cd45dd4c29babf21ffcef09978R127

they should be using `Tracing.endpoint_to_string Bugtool`: https://github.com/xapi-project/xen-api/blob/master/ocaml/libs/tracing/tracing.ml#L54

Reply:
might be easier if we do this in one place. I can create a new PR just for this variant and then we can buidl on that

Reply:
see #5269